File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> arr = new ArrayList<>();
        for (int i : nums) {
            arr.add(i);
        }
        return helper(arr, new ArrayList<>());
    }
    
    public List<List<Integer>> helper(List<Integer> nums, List<Integer> curr) {
        List<List<Integer>> ans = new ArrayList<>();
        if (nums.isEmpty()) {
            ans.add(curr);
            return ans;
        }
        for (int i : nums) {
            List<Integer> a = new ArrayList<>(nums);
            List<Integer> b = new ArrayList<>(curr);
            b.add(i);
            a.remove((Integer) i);
            ans.addAll(helper(a, b));
        }
        return ans;
    }
}

File: 0046-permutations/0046-permutations.swift
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
    return helper(nums, [])
}

func helper(_ nums: [Int], _ curr: [Int]) -> [[Int]] {
    var ans: [[Int]] = []
    if nums.isEmpty {
        let res: [Int] = curr
        ans.append(res)
        return ans
    }
    for i in 0..<nums.count {
        var newNums: [Int] = nums
        var newCurr: [Int] = curr
        newNums.remove(at: i)
        newCurr.append(nums[i])
        ans.append(contentsOf: helper(newNums, newCurr))
    }
    return ans
}



}